src/Microsoft.Identity.Client/Cache/TokenCacheJsonSerializer.cs

```diff
@@ -19,35 +41,40 @@ public TokenCacheJsonSerializer(ITokenCacheAccessor accessor)
 
         public byte[] Serialize()
         {
-            var cache = new CacheRoot();
-            foreach (var t in _accessor.GetAllAccessTokens())
+            var cache = new CacheSerializationContract();
+            foreach (var token in _accessor.GetAllAccessTokens())
             {
-                cache.AccessTokens[t.GetKey().ToString()] = t;
+                cache.AccessTokens[token.GetKey()
+                                    .ToString()] = token;
             }
 
-            foreach (var t in _accessor.GetAllRefreshTokens())
+            foreach (var token in _accessor.GetAllRefreshTokens())
             {
-                cache.RefreshTokens[t.GetKey().ToString()] = t;
+                cache.RefreshTokens[token.GetKey()
+                                     .ToString()] = token;
             }
 
-            foreach (var t in _accessor.GetAllIdTokens())
+            foreach (var token in _accessor.GetAllIdTokens())
             {
-                cache.IdTokens[t.GetKey().ToString()] = t;
+                cache.IdTokens[token.GetKey()
+                                .ToString()] = token;
             }
 
-            foreach (var t in _accessor.GetAllAccounts())
+            foreach (var acct in _accessor.GetAllAccounts())
             {
-                cache.Accounts[t.GetKey().ToString()] = t;
+                cache.Accounts[acct.GetKey()
+                                .ToString()] = acct;
             }
 
-            return cache.ToString().ToByteArray();
+            return cache.ToJsonString()
+                        .ToByteArray();
         }
 
         public void Deserialize(byte[] bytes)
         {
             _accessor.Clear();
 
-            var cache = CacheRoot.FromJsonString(CoreHelpers.ByteArrayToString(bytes));
+            var cache = CacheSerializationContract.FromJsonString(CoreHelpers.ByteArrayToString(bytes));
```

Comment:
Previous versions of the cache allowed the blob to be null. So I would think that customers who don't care about V2 -> V2 would like to just update their existing code by changing Serialize=>SerializeV3 ...

Customers typically would like to update their code this way:
```CSharp
        public static void BeforeAccessNotification(TokenCacheNotificationArgs args)
        {
            lock (FileLock)
            {
                args.TokenCache.DeserializeV3(File.Exists(CacheFilePath)
                    ? ProtectedData.Unprotect(File.ReadAllBytes(CacheFilePath),
                                              null,
                                              DataProtectionScope.CurrentUser)
                    : null);
            }
        }
```
and today this throws in `CacheSerializationContract.FromJsonString()` if json is null (of course)

**My recommendation** would be that if `bytes` is null we do nothing.

Reply:
BTW do customer need to apply a lock in V3? (not sure what this is doing, bTW)? Or is it safe like in V2?

Reply:
re locking:  we lock internally so the customer shouldn't have lock around us.  but they may need to lock the file they're writing to if they're doing multi-app access.

i've added the code to ensure we don't fail if bytes is null or empty

It's a bit more on migration.  
* If they are their own app and don't care about sharing their cache, they will need to deserialize (once) from v2 and then they can serialize/deserialize to v3.  the v3 serializer isn't "smart" to do the read detection of the old format for them.
* If they care about sharing their cache, they will need to build their own logic either staying on existing serialize/deserialize (current dictionary format) or whether and when they can move to v3

Reply:
We _could_ look at making this a bit smarter (e.g. deserializev3 could try the new format and if that fails, try to fallback to the old format).  But the developer will need explicit control over serialize/serializev3 to maintain compat in the way that works for them.  Perhaps for consistency in the "i need to do both" case, we could have a SerializeV3 that returns _both_ formats that are serialized within the same lock (so they're guaranteed consistent)...